// @subsquid/util-internal-validation@0.8.0/lib/dsl.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.object = object;
exports.record = record;
exports.taggedUnion = taggedUnion;
exports.keyTaggedUnion = keyTaggedUnion;
exports.tuple = tuple;
exports.array = array;
exports.option = option;
exports.nullable = nullable;
exports.withSentinel = withSentinel;
exports.withDefault = withDefault;
exports.ref = ref;
exports.oneOf = oneOf;
exports.constant = constant;
exports.cast = cast;
exports.assertValidity = assertValidity;
const array_1 = require("./composite/array");
const constant_1 = require("./composite/constant");
const default_1 = require("./composite/default");
const key_tagged_union_1 = require("./composite/key-tagged-union");
const nullable_1 = require("./composite/nullable");
const object_1 = require("./composite/object");
const one_of_1 = require("./composite/one-of");
const option_1 = require("./composite/option");
const record_1 = require("./composite/record");
const ref_1 = require("./composite/ref");
const sentinel_1 = require("./composite/sentinel");
const tagged_union_1 = require("./composite/tagged-union");
const tuple_1 = require("./composite/tuple");
const error_1 = require("./error");
function object(props) {
    let presentProps = {};
    for (let key in props) {
        let v = props[key];
        if (v) {
            presentProps[key] = v;
        }
    }
    return new object_1.ObjectValidator(presentProps);
}
function record(key, value) {
    return new record_1.RecordValidator(key, value);
}
function taggedUnion(field, variants) {
    return new tagged_union_1.TaggedUnion(field, variants);
}
function keyTaggedUnion(variants) {
    return new key_tagged_union_1.KeyTaggedUnionValidator(variants);
}
function tuple(...tuple) {
    return new tuple_1.TupleValidator(tuple);
}
function array(item) {
    return new array_1.ArrayValidator(item);
}
function option(item) {
    return new option_1.OptionValidator(item);
}
function nullable(item) {
    return new nullable_1.NullableValidator(item);
}
function withSentinel(label, value, validator) {
    return new sentinel_1.Sentinel(label, value, validator);
}
function withDefault(value, validator) {
    return new default_1.Default(value, validator);
}
function ref(get) {
    return new ref_1.RefValidator(get);
}
function oneOf(patterns) {
    return new one_of_1.OneOfValidator(patterns);
}
function constant(value, equals) {
    return new constant_1.ConstantValidator(value, equals);
}
function cast(validator, value) {
    let result = validator.cast(value);
    if (result instanceof error_1.ValidationFailure)
        throw new error_1.DataValidationError(result.toString());
    return result;
}
function assertValidity(validator, value) {
    let err = validator.validate(value);
    if (err)
        throw new error_1.DataValidationError(err.toString());
}
